Don't use an additional buffer in the arithmetic conversions

gennaroprota opened the pull request:
In the case of the floating point conversions, this effectively avoids a copy of the buffer contents. In the case of the integer conversions, it doesn't eliminate the copy, but still removes the extra buffer.

gennaroprota committed:
Don't use an additional buffer in the arithmetic conversions

In the case of the floating point conversions, this effectively avoids a
copy of the buffer contents. In the case of the integer conversions, it
doesn't eliminate the copy, but still removes the extra buffer.

This fixes issue #65.

Flamefire commented:
Maybe merge #64 first to let the full CI run, I disabled C++26 testing just now to let it pass until #66 is resolved

Flamefire commented on include/boost/static_string/static_string.hpp:
```diff
@@ -787,6 +793,32 @@ find_first_of(
   return last;
 }
 
+template<std::size_t N, typename Integer>
```
Why are those declarations required **after** the definitions above?

gennaroprota replied:
Yes, those are not needed, thanks.

Flamefire commented on include/boost/static_string/static_string.hpp:
```diff
@@ -925,6 +957,37 @@ class basic_static_string
   template<std::size_t, class, class>
   friend class basic_static_string;
+
+  template<std::size_t P, typename Integer>
```
They don't use any private properties, do they? So no need for friends

gennaroprota replied:
They use `set_size()`, which is private in `basic_static_string` due to private inheritance.

ashtum replied:
Adding all these friend functions for to_string implementations feels odd. Ideally, users should be able to create their own similar utility functions with comparable performance, but they can't make their functions friends of `static_string`. Would adding a `resize_and_overwrite` member function remove the need for those friends?

gennaroprota replied:
OK. I've added `resize_and_overwrite()` and implemented the arithmetic conversions in terms of it. Please have a look.

ashtum replied:
Ah, it was just a suggestion, but you've already implemented it, and it seems to work.
@pdimov, does adding `resize_and_overwrite()` to `static_string` make sense? Context: Gennaro wants to rewrite the `to_static_string` functions and eliminate the use of a temporary buffer:
https://github.com/boostorg/static_string/blob/0d255f74386a895df08acda33333b17e806fdc6d/include/boost/static_string/static_string.hpp#L552-L556
This enables RVO and produces more efficient assembly: https://godbolt.org/z/bbesGEPj8

pdimov replied:
I suppose it does.

Flamefire commented on include/boost/static_string/static_string.hpp:
```diff
@@ -549,11 +549,15 @@ inline
-  char buffer[N];
-  const auto digits_end = std::end(buffer);
+  static_string<N> result;
+  result.set_size(N);
```
This is redundant here, isn't it?

```suggestion
```

gennaroprota replied:
The first `set_size()` is not redundant with the constructor, because the constructor creates a string with size zero. I could avoid it by writing:

```
const auto digits_end = result.begin() + N;
```

but I don't see any way around the second `set_size()`.

Flamefire replied:
Ah yes, I missed the use of `result.end()`, so it is fine then although a bit confusing to some (like me)

To me the most readable approach would be `const auto digits_end = result.data() + N;` as it highlights the buffer use and mirrors the use in the `move` below.

Flamefire commented on include/boost/static_string/static_string.hpp:
```diff
@@ -562,11 +566,15 @@ inline
-  wchar_t buffer[N];
-  const auto digits_end = std::end(buffer);
+  static_wstring<N> result;
+  result.set_size(N);
```
See above
```suggestion
```

Flamefire commented on include/boost/static_string/static_string.hpp:
```diff
@@ -562,11 +566,15 @@ inline
-  const auto digits_end = std::end(buffer);
+  static_wstring<N> result;
+  result.set_size(N);
+  const auto digits_end = result.end();
```
```suggestion
  const auto digits_end = result.data() + N;
```

ashtum commented:
> this effectively avoids a copy of the buffer contents. In the case of the integer conversions, it doesn't eliminate the copy, but still removes the extra buffer.

Nice, it also enables RVO and produces more efficient assembly because it skips the call to the constructor that is not `noexcept`: https://godbolt.org/z/bbesGEPj8

gennaroprota committed:
Add a resize_and_overwrite() member

Reason: This is in preparation of the next commit. See its commit
message.

ashtum commented on include/boost/static_string/static_string.hpp:
```diff
@@ -6701,6 +6731,33 @@ resize(size_type n, value_type c)
+  Operation op)
+{
+  if (n > max_size()) {
+    detail::throw_exception<std::length_error>("n > max_size() in resize_and_overwrite()");
```
I believe the message "n > max_size()" (identical to what's thrown in resize()) is sufficient. there's no need to proliferate string literals.

gennaroprota replied:
Is it sufficient? It gives no clue to the user about where the problem occurred.

ashtum replied:
I believe `detail::throw_exception` uses `BOOST_THROW_EXCEPTION` under the hood, which would also print the source location. But generally, the exception message doesn't need to include the current function name, "n > max_size()" is descriptive enough for this kind of error message in the code.

pdimov replied:
BOOST_THROW_EXCEPTION, if used inside detail::throw_exception, would emit the source location of detail::throw_exception.

You probably want boost::throw_with_location here, or the two argument overload of boost::throw_exception, so that the source location of the throw is captured, not that of detail::throw_exception.

ashtum replied:
@gennaroprota, it might be more convenient to create an issue for this and address it in a separate PR once you’re done with the current one.

gennaroprota replied:
What if `BOOST_STATIC_STRING_STANDALONE` is defined?

pdimov replied:
If this configuration is to still be supported (is it?) you'd probably need to define BOOST_STATIC_STRING_CURRENT_LOCATION and then use `detail::throw_exception( E(...), BOOST_STATIC_STRING_CURRENT_LOCATION )` instead of `boost::throw_exception( E(...), BOOST_CURRENT_LOCATION )`.

Something better might also be possible, I'm not sure.

gennaroprota replied:
> If this configuration is to still be supported (is it?)

I'd be happy to remove the support. The less conditional compilation, the better.

pdimov replied:
We should probably ask on the list.

ashtum commented on test/static_string.cpp:
```diff
@@ -7368,6 +7368,68 @@ testResize()
 }
 
+void
+testResizeAndOverwrite()
```
It might be better to split this PR into two commits: the first adds the `resize_and_overwrite` function, and the second improves the performance of the `to_string` implementations.

Flamefire commented on include/boost/static_string/static_string.hpp:
```diff
@@ -3751,6 +3763,24 @@ class basic_static_string
+  /**
+      Resize the string and overwrite its contents.
+
+      Resizes the string to contain `n` characters, and uses the
```
I'm not sure that is useful, the size passed here doesn't really matter, does it? Maybe a `write_and_resize(op)` is more useful? Or is there any realistic use case for this param being less than N?

pdimov replied:
That's how the function is specified for std::string.

For static_string specifically the only thing the passed size does is throw if it's too big.

Flamefire replied:
Ok, makes sense

gennaroprota committed:
Implement the arithmetic conversions in terms of resize_and_overwrite()

Reason: Performing the conversions without accessing private members,
providing a model for users to implement their own with comparable
efficiency.
